src/useSprings.js

```diff
@@ -13,52 +13,48 @@ export const useSprings = (length, props) => {
   const isFn = is.fun(props)
 
   // The controller maintains the animation values, starts and stops animations
-  const [controllers, ref] = useMemo(() => {
-    // Remove old controllers
-    if (ctrl.current) {
-      ctrl.current.map(c => c.destroy())
-      ctrl.current = undefined
-    }
+  const [controllers, setProps, ref, api] = useMemo(() => {
     let ref
     return [
+      // Recreate the controllers whenever `length` changes
       new Array(length).fill().map((_, i) => {
-        const ctrl = new Ctrl()
-        const newProps = isFn ? callProp(props, i, ctrl) : props[i]
+        const c = new Ctrl()
+        const newProps = isFn ? callProp(props, i, c) : props[i]
         if (i === 0) ref = newProps.ref
-        return ctrl.update(newProps)
+        return c.update(newProps)
       }),
+      // This updates the controllers with new props
+      props =>
+        ctrl.current.forEach((c, i) => {
+          c.update(is.fun(props) ? callProp(props, i, c) : props[i])
+          if (!ref) c.start()
+        }),
+      // The imperative API is accessed via ref
       ref,
+      ref && {
+        start: () =>
+          Promise.all(ctrl.current.map(c => new Promise(r => c.start(r)))),
+        stop: finished => ctrl.current.forEach(c => c.stop(finished)),
+        get controllers() {
+          return ctrl.current
+        },
+      },
     ]
   }, [length])
 
-  ctrl.current = controllers
```

Comment:
I moved this into the `useEffect` call that runs on every render, since you're not supposed to mutate stuff during render ([see here](https://twitter.com/dan_abramov/status/1102182683885617159)).